### PR TITLE
AI Chat: per-Conversation model selection

### PR DESCRIPTION
**Pull request by petemill**
Resolves https://github.com/brave/brave-browser/issues/32737

image

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -355,7 +355,7 @@
   },
   "dependencies": {
     "@brave/brave-ui": "0.40.4",
-    "@brave/leo": "github:brave/leo#0e93ff5",
+    "@brave/leo": "github:brave/leo#f462e8c8680f16b8d1de058f23ccb76392c71fe9",
```
Comment (petemill) [inline package.json:358]: We'll update this in a separate PR and I'll remove this before merge

File: components/ai_chat/browser/models.cc [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+#include "brave/components/ai_chat/browser/models.h"
+
+#include "brave/components/ai_chat/common/mojom/ai_chat.mojom.h"
+
+namespace ai_chat {
+
+constexpr char kModelsDefaultKey[] = "chat-default";
+
+const base::flat_map<base::StringPiece, mojom::Model> kAllModels = {
```
Comment (nullhook) [inline components/ai_chat/browser/models.cc:16]: Can this be constructed at compile-time? i.e `base::MakeFlatMap`

File: components/ai_chat/browser/models.cc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    //   "code-llama-13b",
+    //   mojom::ModelCategory::CODING,
+    //   false}},
+    // {"code-llama-expanded", {"code-llama-expanded",
+    //   "code-llama-70b",
+    //   mojom::ModelCategory::CODING,
+    //   true}},
+};
+
+const std::vector<base::StringPiece> kAllModelKeysDisplayOrder = {
```
Comment (nullhook) [inline components/ai_chat/browser/models.cc:36]: Can this a be a compile-time flat set?

File: components/ai_chat/browser/models.cc (hunk 2 above)
Comment (petemill) [reply components/ai_chat/browser/models.cc:16]: not for a mojom struct AFAIK unless you know of a way

File: components/ai_chat/browser/models.cc (hunk 3 above)
Comment (petemill) [reply components/ai_chat/browser/models.cc:36]: no, it needs to be in order so it needs to be a vector

File: components/ai_chat/browser/models.cc (hunk 2 above)
Comment (nullhook) [reply components/ai_chat/browser/models.cc:16]: ah, I forget that Mojo types are not constexpr constructible. Their ctors need to be marked as constexpr but then ::New() methods will probably never be constexpr.

File: components/resources/ai_chat_ui_strings.grdp [hunk 4]
```diff
@@ -48,4 +48,31 @@
...
   </message>
+  <message name="IDS_CHAT_UI_INTRO_MESSAGE_CHAT_DEFAULT" desc="AI Chat intro message for the default model">
+    I'm here to help. What can I assist you with today?
+  </message>
+  <message name="IDS_CHAT_UI_INTRO_MESSAGE_CHAT_LEO_EXPANDED" desc="AI Chat intro message for the expanded model">
+    I have a vast base of knowledge and a large memory able to help with more complex challenges.
+  </message>
+  <message name="IDS_CHAT_UI_INTRO_MESSAGE_CHAT_LEO_CLAUDE_INSTANT" desc="AI Chat intro message for the Claude Instant model">
+    Hi! My name is Claude. I was created by Anthropic to be helpful, harmless, and honest.
+  </message>
```
Comment (mkarolin) [inline components/resources/ai_chat_ui_strings.grdp:59]: Are these used?

File: components/resources/ai_chat_ui_strings.grdp [hunk 5]
```diff
@@ -48,4 +48,31 @@
...
+    <ph name="MODEL_NAME">$1<ex>llama2-13b</ex></ph> by <ph name="COMPANY_NAME">$2<ex>Meta</ex></ph>
+  </message>
+  <message name="IDS_CHAT_UI_MODEL_CATEGORY_CHAT" desc="Category name for Chat">
+    Chat
+  </message>
+  <message name="IDS_CHAT_UI_MENU_NEW_CHAT" desc="Menu item for new chat">
+    New chat
+  </message>
+  <message name="IDS_CHAT_UI_MENU_SUGGESTED_QUESTIONS" desc="Menu item for whether to enable suggested questions during chat conversations">
+    Suggested questions
```
Comment (mkarolin) [inline components/resources/ai_chat_ui_strings.grdp:70]: Not sure about this one. Personally, either `Show suggested questions` or `Suggest questions` work better to me.

File: components/resources/ai_chat_ui_strings.grdp (hunk 4 above)
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:59]: yes, they are used in the f/e

Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:59]:
```cpp
      {"introMessage-chat-default", IDS_CHAT_UI_INTRO_MESSAGE_CHAT_DEFAULT},
      {"introMessage-chat-leo-expanded",
       IDS_CHAT_UI_INTRO_MESSAGE_CHAT_LEO_EXPANDED},
      {"introMessage-chat-claude-instant",
       IDS_CHAT_UI_INTRO_MESSAGE_CHAT_LEO_CLAUDE_INSTANT},
```

File: components/resources/ai_chat_ui_strings.grdp (hunk 5 above)
Comment (petemill) [reply components/resources/ai_chat_ui_strings.grdp:70]: Agreed. We use `Enable suggested questions?` in other places, but here if we're going for the short version then `Suggest questions` seems ok @aguscruiz

File: components/resources/ai_chat_ui_strings.grdp (hunk 4 above)
Comment (mkarolin) [reply components/resources/ai_chat_ui_strings.grdp:59]: I saw them being mapped, but I didn't see the mapped strings (like `introMessage-chat-default`) being used elsewhere.

**Review by nullhook — APPROVED** [review]
lgtm

**Review by zenparsing — APPROVED** [review]
Rewards 👍

**Review by mkarolin — APPROVED** [review]
`strings`++

**Comment by petemill** [comment]
follow-up: https://github.com/brave/brave-browser/issues/33308

**Pull request merged**